core/src/main/scala/org/apache/spark/rdd/ReliableCheckpointRDD.scala

```diff
@@ -82,14 +83,28 @@ private[spark] class ReliableCheckpointRDD[T: ClassTag](
     Array.tabulate(inputFiles.length)(i => new CheckpointRDDPartition(i))
   }
 
+  // Cache of preferred locations of checkpointed files.
+  private[spark] val cachedPreferredLocations: mutable.HashMap[Int, Seq[String]] =
+    mutable.HashMap.empty
```

Comment:
If there is a side-effect, this will be the usual suspect. I'm wondering if we need a Caching policies via GuavaCache.

Reply:
Once checkpointed to DFS, I think the locations of checkpointed files should not be changed during job execution. This cache is not unlimited in size, as the number RDD files are fixed. But, of course, as this is core, we should be more careful.

Reply:
What I meant was the size constraint instead of the time expiration. How large do you expect the size of this cache?

Reply:
getPreferredLocations only takes the hostnames from one BlockLocation for each partition. The number of hostnames depends on the block replicas on DFS. So if block replicas is 3, we cache 3 hostname strings for each partition.

Reply:
The following assumption sounds weak to me. HDFS NN also returns the locations based on the data nodes situation and data nodes can die any time. If a data node with replica dies, HDFS is going to recover it and returns different locations (including existing ones). This PR seems to imply we have always a corrupted set of host names.
> I think the locations of checkpointed files should not be changed during job execution.

Reply:
This is good point. Although the cached preferred locations can be possibly outdated, I think it only implies that the location to launch tasks could be not best options.

For example, if the cached preferred loc is [hostA, hostB, hostC] for partition 1. When hostA dies, if Spark executor also dies there, we will consider other preferred locations hostB and hostC.

If only data node dies at hostA but Spark executor still works there, Spark could still choose hostA to launch the task. In this case, hostA is just not best option.

Yeah, I think the assumption is not so strong, the locations are just relatively stable during job execution. So this is a config for users to enable or not, not a feature always turn on.

Reply:
If ALS runs more longer, this PR will cause Job failure.
Let's say Spark Job start with cache (A1,A2,A3) and we do the rolling restart HDFS.

HDFS can do the followings.
1. A1, A2, A3
2.     , A2, A3   (<= A1 is shutdown)
3. B1, A2, A3   (<= HDFS create a new replica B1)
4. B1,     , A3
5. B1, B2, A3
6. B1, B2, 
7. B1, B2, B3

Actually, the above is HDFS design. After this PR, Spark job starts to fail at step 6.

Reply:
I think you meant when at step 6, all replicas are placed different hosts in our cached, right?

Since A1 is shutdown, Spark won't launch task on it. Say Spark launches task at A2 or A3, the task still can access block from B1 or B2. Could you explain why it will fail?

Reply:
?
When the job starts, this PR will cache `A1, A2, A3` and will not update it due to the following, right?
```
if (cachePreferrredLoc && cachedPreferredLocations.contains(split.index)) {
```

When do you think Spark get `Bx`?

For example, when the long running task is assigned to A2, then A2 die. Then, Spark will retry with A3 and will run. Then, A3 dies. At the 3rd try, Spark only knows `A1, A2, A3` not `B1, B2, B3? Did I understand correctly?

Reply:
Ur, Spark not only chooses hosts from preferred locations. It is just a preferred list, Spark still can launch task if the list of hosts are unavailable.

If a preferred location is present, the task will be added into pending task set for the host. But it is also added into a set of all tasks.

https://github.com/apache/spark/blob/b29829e2abdebdf6fa9dd0a4a4cf4c9d676ee82d/core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala#L238-L251

The task in the set of all tasks can still be dequeued and scheduled:

https://github.com/apache/spark/blob/b29829e2abdebdf6fa9dd0a4a4cf4c9d676ee82d/core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala#L358-L362

Reply:
Oh, Got it. I missed that point. This is only for task scheduling instead of data access.
So, the problem is narrowed down to that Spark job will lose data locality completely, isn't it?

Reply:
Yes, it is the worst case the cached preferred locations are outdated at all.

Reply:
Then, we have two options.
1. Document the limitation in the PR description and `config` document and continue in this way.
2. Using Guava cache with timeout policy to mitigate the possibility.

For the size, it looks okay according to your estimation. How do you want to proceed this?

Reply:
Thanks @dongjoon-hyun for your review and suggestion! 

I think Guava cache can enable more control on caching behavior like expire time. I would use option 2 for now.